engines/scumm/he/moonbase/map_spiff.cpp

```diff
@@ -0,0 +1,623 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/debug.h"
+
+#include "scumm/he/moonbase/map_spiff.h"
+
+namespace Scumm {
+
+SpiffGenerator::SpiffGenerator(int seed) {
+	_seed = seed;
+}
+
+SpiffGenerator::~SpiffGenerator() {
+}
+
+MapFile *SpiffGenerator::generateMap(int water, int tileset, int mapSize, int energy, int terrain) {
+	totalMapSizeG = mapSize;
+	energyAmountG = (2 + energy) * totalMapSizeG * totalMapSizeG;
+
+	islandsFlagG = pickFrom2(0, 1, water - 4, (water >= 5)); // 1 is large islands, 2 is small
+	if (islandsFlagG) {
+		water -= 3;
+		energyAmountG = (int)(energyAmountG * (5 - islandsFlagG) / 6); // *2/3 or *1/2
+	}
+
+	waterAmountG = 4 * water;
+	cliffAmountG = 1 << terrain;
+	advancedMirrorOK_G = ((terrain > 1) && (water < 6)) || islandsFlagG;
+	terrainSeedFlagG = 2 * water - terrain;
+
+	int n = (int)(energyAmountG / 2700);
+	if (n > 12) {
+		n = 12;
+	}
+	if (n < 1) {
+		n = 1;
+	}
+	numPoolsG = spiffRand((int)(energyAmountG / 4000) + 1, n);
+	if (numPoolsG > 12) {
+		numPoolsG = 12;
+	}
+
+	generate();
+
+	// Populate MIF for map data generation:
+	MIF mif = MIF();
+
+	int levelMap[MAXELEVVAL];
+	levelMap[HIGH] = 2;
+	levelMap[MEDIUM] = 1;
+	levelMap[LOW] = 0;
+
+	mif.mapType = tileset;
+	Common::sprintf_s(mif.name, "Spiff %04X", (uint16)_seed);
+
+	mif.dimension = totalMapSizeG;
+
+	int y;
+	int x;
+	char t;
```

Comment:
It is better to write `byte`, because `char` could have a sign or even be multibyte.